### PR TITLE
Add support for atmega256rfr2 / Pinoccio scout

### PR DESCRIPTION
**Pull request by matthijskooijman**
As proposed in #17, this adds support for using Pinoccio Scout board as programmer and target. Additionally, it improves the message shown when a device is supported, but it bootloader is not enabled at compiletime.

Please review these commits, but hold off on merging them. The bootloader included is not final yet. If no problems are found, this will become the final version, but it needs a bit more testing first.

### CODE REVIEW

**Comment by matthijskooijman** [comment]
And here's a proposal for documentation:


Using a Atmega256rfr2 / Pinoccio Scout as a programmer or target
================================================================
Since *TODO DATE* this sketch also supports Atmega256rfr2-based boards, like the Pinoccio Scout. This board can be used as the programmer or target board.

When using the Pinoccio Scout board, take special care about the operating voltage. The Scout runs at 3.3V, so you cannot use it together with a board running at 5V (like most Arduinos) without using a voltage converter. Using two Scouts together, or a Scout and a 3.3V Arduino (like the Fio or an Arduino Pro at 3.3V) should be possible.

Since the Pinoccio Scout does not have a pin 10 available, the SS pin is used instead. On the target device, the reset (RST) pin is used as normal.

In the image below, the target is powered by connecting the VUSB pins together. Alternatively, you can just power the target board through USB as well (but still connect GND in this case). Remember that you cannot power a pinoccio board through the 3V3 pin, which is output only (and not enabled when running this sketch).

Note that the Atmega256rfr2 bootloader is _not_ enabled by default. To enable it, modify the Atmega_Board_Programmer.ino before uploading the sketch.

drawing

File: Atmega_Board_Programmer/Atmega_Board_Programmer.ino [hunk 1]
```diff
@@ -500,7 +529,10 @@ void writeBootloader ()
 
   if (signatures [foundSig].bootloader == 0)
     {
-    Serial.println (F("No bootloader support for this device."));
+    if (signatures [foundSig].loaderStart != 0)
+      Serial.println (F("Bootloader for this device is disabled, edit " __FILE__ " to enable it."));
+    else
+      Serial.println (F("No bootloader support for this device."));
     return;
```
Comment (NicoHood) [inline Atmega_Board_Programmer/Atmega_Board_Programmer.ino]: I like this part in the patch.

**Comment by nickgammon** [comment]
```
   if (signatures [foundSig].bootloader == 0)
     {
-    Serial.println (F("No bootloader support for this device."));
+    if (signatures [foundSig].loaderStart != 0)
```

How can signatures [foundSig].loaderStart possibly be non-zero at this point? We already know it is zero.

**Comment by matthijskooijman** [comment]
> How can signatures [foundSig].loaderStart possibly be non-zero at this point? We already know it is zero.

No, we know that `bootloader` is zero. When a bootloader is disabled through the list of macros at the top, the `bootloader` and (`bootloaderSize` or whatever it is called) are zero, but the other fields keep their values. So testing for `loaderStart` allows checking if there is a bootloader for the device, which is disabled if `bootloader` is zero.

Admittedly, this is a bit messy, so perhaps some refactoring would be appropriate here, but I just wanted to go for the least invasive way to implement this first.

**Comment by nickgammon** [comment]
Oh I see, I misread it. :)

**Comment by nickgammon** [comment]
Updated version number, merged pull request.

https://github.com/nickgammon/arduino_sketches/commit/fda72394567d8c44e3e4c045db04403ef61d7e69

**Comment by nickgammon** [comment]
Web page http://www.gammon.com.au/bootloader updated.

**Comment by matthijskooijman** [comment]
Hmm, I presume you didn't see this part of my initial message?

> Please review these commits, but hold off on merging them. The bootloader included is not final yet. If no problems are found, this will become the final version, but it needs a bit more testing first.

We haven't completed testing yet, but let's assume that this will indeed be the final version. If not, I'll just bump the bootlloader version number and submit another PR :-)

Thanks!

**Comment by nickgammon** [comment]
No, sorry. Oh well, just submit a change if required. I waited 4 days anyway. :) Then I thought you might be getting impatient. If I'd re-read your first post I need not have worried.

**Pull request merged**